Static graph restore failure when referencing unrestorable project

### NuGet Product Used

dotnet.exe, MSBuild.exe

### Product Version

.NET SDK 7.0.100; VS 17.5 internal preview

### Worked before?

Not a regression

### Impact

It's more difficult to complete my work

### Repro Steps & Context

Given a project that has a `ProjectReference` to a project that doesn't import `NuGet.targets`, like `Something\Something.csproj` referencing the empty project `asdf.proj` in [GraphRestoreFailsOnUnrestorableReference.zip](https://github.com/NuGet/Home/files/10239519/GraphRestoreFailsOnUnrestorableReference.zip), graph-based restore can fail when non-graph-based restore succeeds.

```sh-session
❯ dotnet restore  -p:RestoreUseStaticGraphEvaluation=true .\Something\
  Determining projects to restore...
S:\play\alkestj\Something\Something.csproj : error NU1105: Unable to find project information for 'S:\play\alkestj\asdf.proj'. If you are using Visual Studio, this may be because the project is unloaded or not part of the current solution so run a restore from the command-line. Otherwise, the project file may be invalid or missing 
targets required for restore.
  Failed to restore S:\play\alkestj\Something\Something.csproj (in 54 ms).
```

Looking at the dgspec, non-graph restore elides the `projectReferences` entry for the reference to the non-restorable project, while graph restore includes it.

```diff
diff --git "a/.\\Something\\obj\\nongraph_Something.csproj.nuget.dgspec.json" "b/.\\Something\\obj\\Something.csproj.nuget.dgspec.json"
index 3468b15..e9b26cf 100644
--- "a/.\\Something\\obj\\nongraph_Something.csproj.nuget.dgspec.json"
+++ "b/.\\Something\\obj\\Something.csproj.nuget.dgspec.json"
@@ -32,7 +32,11 @@
         "frameworks": {
           "net7.0": {
             "targetAlias": "net7.0",
-            "projectReferences": {}
+            "projectReferences": {
+              "S:\\play\\alkestj\\asdf.proj": {
+                "projectPath": "S:\\play\\alkestj\\asdf.proj"
+              }
+            }
           }
         },
         "warningProperties": {
```

Since there is in fact no info for that project, this leads to a failure later.

### Verbose Logs

_No response_

--- a/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
@@ -706,6 +706,9 @@ namespace NuGet.Build.Tasks.Console
                 // Fix project reference casings to match the original project on case insensitive file systems.
                 MSBuildRestoreUtility.NormalizePathCasings(projectPathLookup, dependencyGraphSpec);
 
+                // Remove references to projects that could not be read by restore.
+                MSBuildRestoreUtility.RemoveMissingProjects(dependencyGraphSpec);
+
                 // Add all entry projects if they support restore.  In most cases this is just a single project but if the entry
                 // project is a solution, then all projects in the solution are added (if they support restore)
                 foreach (var entryPoint in entryProjects)

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
@@ -316,6 +316,7 @@ namespace NuGet.Commands
             {
                 foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
+                    // Loop through the items in reverse order so items can be removed from the collection safely
                     for (int i = framework.ProjectReferences.Count - 1; i >= 0; i--)
                     {
                         if (!existingProjects.Contains(framework.ProjectReferences[i].ProjectPath))

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
@@ -316,11 +316,11 @@ namespace NuGet.Commands
             {
                 foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
-                    foreach (var projectReference in framework.ProjectReferences)
+                    for (int i = framework.ProjectReferences.Count - 1; i >= 0; i--)
                     {
-                        if (!existingProjects.Contains(projectReference.ProjectPath))
+                        if (!existingProjects.Contains(framework.ProjectReferences[i].ProjectPath))
                         {
-                            framework.ProjectReferences.Remove(projectReference);
+                            framework.ProjectReferences.Remove(framework.ProjectReferences[i]);
                         }
                     }
                 }

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
@@ -316,7 +316,7 @@ namespace NuGet.Commands
             {
                 foreach (var framework in project.RestoreMetadata.TargetFrameworks)
                 {
-                    foreach (var projectReference in framework.ProjectReferences.ToArray())
+                    foreach (var projectReference in framework.ProjectReferences)
                     {
                         if (!existingProjects.Contains(projectReference.ProjectPath))
                         {

--- a/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
+++ b/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
@@ -810,7 +810,7 @@ $@"<?xml version=""1.0"" encoding=""utf-8""?>
         [PlatformTheory(Platform.Windows)]
         [InlineData(true)]
         [InlineData(false)]
-        public void MsbuildRestore_WithMissingProjectReferences_HandlesProjectReferencesToUnsupportedProjects(bool restoreUseStaticGraphEvaluation)
+        public async Task MsbuildRestore_WithMissingProjectReferences_HandlesProjectReferencesToUnsupportedProjects(bool restoreUseStaticGraphEvaluation)
         {
             // Arrange
             using (var pathContext = new SimpleTestPathContext())
@@ -818,16 +818,13 @@ $@"<?xml version=""1.0"" encoding=""utf-8""?>
                 // Set up solution, project, and packages
                 var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
 
-                NuGetFramework targetFramework = NuGetFramework.Parse("net461");
+                var net461 = NuGetFramework.Parse("net472");
 
-                var projectA = new SimpleTestProjectContext("ProjectA", ProjectStyle.PackageReference, pathContext.SolutionRoot)
-                {
-                    ToolingVersion15 = true
-                };
+                var projectA = new SimpleTestProjectContext("a", ProjectStyle.PackageReference, pathContext.SolutionRoot);
 
                 var projectB = new SimpleTestProjectContext("b", ProjectStyle.PackageReference, pathContext.SolutionRoot);
 
-                var projectAFrameworkContext = new SimpleTestProjectFrameworkContext(targetFramework);
+                var projectAFrameworkContext = new SimpleTestProjectFrameworkContext(net461);
 
                 projectAFrameworkContext.ProjectReferences.Add(projectB);
 
@@ -845,15 +842,15 @@ $@"<?xml version=""1.0"" encoding=""utf-8""?>
                 solution.Create(pathContext.SolutionRoot);
 
                 File.WriteAllText(
-                    projectB.ProjectPath,
-                    @"<Project />");
+                   projectB.ProjectPath,
+                   @"<Project />");
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(pathContext.PackageSource, packageX);
 
                 var result = _msbuildFixture.RunMsBuild(pathContext.WorkingDirectory, $"/t:restore /p:RestoreUseStaticGraphEvaluation={restoreUseStaticGraphEvaluation} {projectA.ProjectPath}", ignoreExitCode: true);
 
                 // Assert
                 result.ExitCode.Should().Be(0, result.AllOutput);
-
-                //result.AllOutput.Should().Contain($"error MSB4025: The project file could not be loaded. Could not find file '{projectB.ProjectPath}'");
             }
         }
 

--- a/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
+++ b/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
@@ -810,6 +810,56 @@ $@"<?xml version=""1.0"" encoding=""utf-8""?>
         [PlatformTheory(Platform.Windows)]
         [InlineData(true)]
         [InlineData(false)]
+        public void MsbuildRestore_WithMissingProjectReferences_HandlesProjectReferencesToUnsupportedProjects(bool restoreUseStaticGraphEvaluation)
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Set up solution, project, and packages
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+
+                NuGetFramework targetFramework = NuGetFramework.Parse("net461");
+
+                var projectA = new SimpleTestProjectContext("ProjectA", ProjectStyle.PackageReference, pathContext.SolutionRoot)
+                {
+                    ToolingVersion15 = true
+                };
+
+                var projectB = new SimpleTestProjectContext("b", ProjectStyle.PackageReference, pathContext.SolutionRoot);
+
+                var projectAFrameworkContext = new SimpleTestProjectFrameworkContext(targetFramework);
+
+                projectAFrameworkContext.ProjectReferences.Add(projectB);
+
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+
+                projectA.Frameworks.Add(projectAFrameworkContext);
+
+                projectA.AddPackageToAllFrameworks(packageX);
+
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+
+                File.WriteAllText(
+                    projectB.ProjectPath,
+                    @"<Project />");
+
+                var result = _msbuildFixture.RunMsBuild(pathContext.WorkingDirectory, $"/t:restore /p:RestoreUseStaticGraphEvaluation={restoreUseStaticGraphEvaluation} {projectA.ProjectPath}", ignoreExitCode: true);
+
+                // Assert
+                result.ExitCode.Should().Be(0, result.AllOutput);
+
+                //result.AllOutput.Should().Contain($"error MSB4025: The project file could not be loaded. Could not find file '{projectB.ProjectPath}'");
+            }
+        }
+
+        [PlatformTheory(Platform.Windows)]
+        [InlineData(true)]
+        [InlineData(false)]
         public async Task MsbuildRestore_WithCPPCliVcxproj_RestoresSuccessfullyWithPackageReference(bool isStaticGraphRestore)
         {
             // Arrange

